kubespawner/spawner.py

```diff
@@ -1809,18 +1836,32 @@ def _expand_user_properties(self, template):
         safe_username = escapism.escape(
             self.user.name, safe=safe_chars, escape_char='-'
         ).lower()
+
+        if self.slug_scheme == "safe":
+            # safe slug scheme escapes _after_ rendering the template
+            username = self.user.name
+            servername = raw_servername
+        else:
+            username = safe_username
+            servername = safe_servername
```

Comment:
Maybe we could also move the code generating `username` and `servername` inside the if, we do not need to compute these steps if we are not using this non `safe` slug.

Reply:
I think they need to be made available no matter what, because it should always be possible to reference `{safe_username}` even though `{username}` only conditionally becomes `safe_username` for example.